src/index/index_construct.c

```diff
@@ -0,0 +1,238 @@
+/*
+* Copyright 2018-2022 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "RG.h"
+#include "index.h"
+#include "../graph/rg_matrix/rg_matrix_iter.h"
+
+#include <assert.h>
+
+// index nodes in an asynchronous manner
+// nodes are being indexed in batchs while the graph's read lock is held
+// to avoid interfering with the DB ongoing operation after each batch of nodes
+// is indexed the graph read lock is released
+// alowing for write queries to be processed
+//
+// it is safe to run a write query which effects the index by either:
+// adding/removing/updating an entity while the index is being populated
+// in the "worst" case we will index that entity twice which is perfectly OK
+static void _Index_PopulateNodeIndex
+(
+	Index *idx,
+	Graph *g
+) {
+	ASSERT(g   != NULL);
+	ASSERT(idx != NULL);
+
+	GrB_Index          rowIdx     = 0;
+	int                indexed    = 0;     // number of entities indexed in current batch
+	int                batch_size = 1000;  // max number of entities to index in one go
+	RG_MatrixTupleIter it         = {0};
+
+	while(true) {
+		// lock graph for reading
+		Graph_AcquireReadLock(g);
+
+		// index state changed, abort indexing
+		// this can happen if for example the following sequance is issued:
+		// 1. CREATE INDEX FOR (n:Person) ON (n.age)
+		// 2. CREATE INDEX FOR (n:Person) ON (n.height)
+		if(idx->pending_changes > 1) {
+			break;
+		}
+
+		// reset number of indexed nodes in batch
+		indexed = 0;
+
+		// fetch label matrix
+		const RG_Matrix m = Graph_GetLabelMatrix(g, idx->label_id);
+		ASSERT(m != NULL);
+
+		//----------------------------------------------------------------------
+		// resume scanning from rowIdx
+		//----------------------------------------------------------------------
+
+		RG_MatrixTupleIter_attach(&it, m);
+		RG_MatrixTupleIter_jump_to_row(&it, rowIdx);
```

Comment:
use RG_MatrixTupleIter_AttachRange

Reply:
you're right, I've caught that, at the moment I'm using:
```c
info = RG_MatrixTupleIter_iterate_range(&it, rowIdx, UINT64_MAX);
```

The usage of `UINT64_MAX` isn't that nice, I would prefer to have a handy function which places the iterator on the first active entry within the range of `rowIdx` ... `end of matrix`